Allow use of other property names than 'encrypted'

cwhatley opened the pull request:
The way this was written, the vault can only be named 'encrypted' and
there can only be one vault per entity. This allows multiple
arbitrarily-named vaults with different keys.

This does change the api for constructing fields from simply field(name)
to fieldInVault(name, vaultName).

cwhatley committed:
Allow use of other property names than 'encrypted'

The way this was written, the vault can only be named 'encrypted' and
there can only be one vault per entity. This allows multiple
arbitrarily-named vaults with different keys.

This does change the api for constructing fields from simply field(name)
to fieldInVault(name, vaultName).

defunctzombie commented:
don't really mind but this could be done in a backwards compatible way :) If you do that, then I am more inclined to accept the PR. Something like maybe tracking the available vaults. Alternatively, maybe have `.vault` return some sort of wrapper which you can then call `.field` on as well as `.vault_field()` With the former giving you a field in the vault and the latter being the actual encrypted vault field.

Anyhow, what is most important to me is either backwards compat or a reworked api that is more consistent and less error prone.

cwhatley committed:
Made API backwards compatible + vault naming opt.

Added vaultField option to constructor.

cwhatley commented:
In that case, I think the simplest thing is to move the name of the vault into the options of the constructor of `EncryptedField` and it can default to 'encrypted'. See last commit.

`var encryptedField = new EncryptedField(Sequelize, key, { vaultField: 'encrypted' })`

Then to define the vault, you'd just ask for `encryptedField.vault()` and to define a field, you'd do `encryptedField.field(name)`

defunctzombie commented:
That seems pretty good to me. I would probably also call the method `.vaultField()` for consistency/clarity. And then in the docs we will just note that it must match the name of the sequelize field in your model (unless you know of a trick or documented way to get the field name from within the data type declaration). Maybe since the vault fields will be from a known named vault, when they are set/unset they could do a model check. Anyhow, not critical since the current code doesn't do anything smart like that but maybe something to think about for the future.

cwhatley committed:
Refactored API.

I feel like this approach is cleaner and more flexible (on the
outside). No worries if you aren't into it and don't want to accept the
PR.

pukhalski commented:
Is it going to be merged? I'm half-way done with my own plugin, while discovered this one. But definitely it needs to be merged.

AndrewAskins commented:
Second this pull request, another super simple option is just to store the name of the vault in a property when the vault is initialized. Then use the stored name to fetch the vault instead of this.encrypted.

defunctzombie commented:
Closing this in favor of https://github.com/defunctzombie/sequelize-encrypted/pull/5 which is similar to the first approach by @cwhatley here with the slight tweak that the encrypted name is figured out from the `.vault()` call. With #5 you can now have multiple vaults in a single sequelize model. Those wanting this feature, please give that a test and let me know.
